import { Request, Response, NextFunction } from "express";
import { validationResult, ValidationChain } from "express-validator";
import ApiError from "@/utils/ApiError";

export const validate = (validations: ValidationChain[]) => {
  return async (req: Request, _res: Response, next: NextFunction) => {
    // Execute all validations
    await Promise.all(validations.map((validation) => validation.run(req)));

    const errors = validationResult(req);

    if (errors.isEmpty()) {
      return next();
    }

    const extractedErrors = errors.array().map((err) => ({
      field: err.type === "field" ? err.path : undefined,
      message: err.msg,
    }));

    throw ApiError.badRequest("Validation error", extractedErrors);
  };
};
